Migrate from logrus to slog

aanm opened the pull request:
Note for reviewers:

I couldn't split this into further commits given the changes in the pkg/logging package required changes everywhere else. :disappointed: 

I would appreciate closer review on the introduction of the multihandler (first commit) and on the second commit the changes done in the pkg/logging as well as the pkg/metrics to detect the presence of duplicate attributes in the slog log messages.

aanm committed:
migrate from logrus to slog

With this commit we migrate the remaining Cilium code base to use slog
and stop relying on logrus.

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
pkg/logging: add multihandler slog.Handler

Add a multihandler slog.Handler so that we can add multiple handlers for
one instance of slog.Logger. This is necessary to keep existing Cilium
functionality that can log one message into multiple logger.

Signed-off-by: André Martins <andre@cilium.io>

joamaki commented on pkg/logging/multihandler.go:
```diff
@@ -0,0 +1,82 @@
+	}
+}
+
+func (i *multiSlogHandler) AddHooks(handlers ...slog.Handler) {
```
Should this be called `AddHandlers`?

aanm replied:
Yep!

joamaki commented on pkg/bgpv1/manager/reconciler/preflight.go:
```diff
@@ -62,7 +62,6 @@ func (r *PreflightReconciler) Cleanup(_ *instance.ServerWithConfig) {}
 func (r *PreflightReconciler) Reconcile(ctx context.Context, p ReconcileParams) error {
 	var (
 		l = r.logger.With(
-			types.ComponentLogField, "PreflightReconciler",
```
I guess this and the other removed attributes were duplicates?

aanm replied:
Yes! This can actually be part of a separate commit.

joamaki commented on pkg/logging/klog_bridge.go:
```diff
@@ -0,0 +1,183 @@
+	},
+}
+
+func initializeKLog() error {
```
Might make sense to avoid `DefaultSlogLogger` in cases where we can, e.g. here we can pass it down from `SetupLogging`

devodev commented on pkg/logging/multihandler.go:
```diff
@@ -0,0 +1,82 @@
+	defer i.mu.RUnlock()
+	for _, h := range i.handlers {
+		return h.Enabled(ctx, level)
+	}
```
```suggestion
	for _, h := range i.handlers {
		if h.Enabled(ctx, level) {
		    return true
		}
	}
```
I think this is what we want right? Otherwise we always return the first handler state.

aanm committed:
add CI detector for duplicated slog attributes

Add a detector in the CI pipeline to catch duplicate attributes in
slog logs. This prevents ambiguous or conflicting keys in structured logs,
ensuring cleaner and more reliable log output. It also helps maintain
consistency and simplifies downstream log processing and analysis.

Signed-off-by: André Martins <andre@cilium.io>

sayboras reviewed (approved):
Looks good for gateway-api and envoy packages.

giorio94 commented on pkg/lock/lock_debug.go:
Nit. These changes do not belong to this commit.

aanm replied:
@giorio94 they do because pkg/logging/multihandler.go creates a dep-cycle with pkg/lock.

giorio94 replied:
`SetLogger` is never called in that commit though. Ideally that change should be moved to its own self-contained commit for clarity, but that's definitely a nit.

joestringer replied:
Strictly speaking yes this resolves the dependency loop, but also this commit breaks logging for `lockdebug` builds because `SetLogger()` is not called until a later commit. That said I don't know that there's a super clean way to resolve that nit other than by merging this into the later commit, but it's nice to be able to focus on review for this commit independently.

giorio94 commented on pkg/lock/lock_fast.go:
Nit. These changes do not belong to this commit.

aanm replied:
same answer as before.

joestringer commented on pkg/logging/multihandler.go:
```diff
@@ -0,0 +1,84 @@
+	"github.com/cilium/cilium/pkg/lock"
+)
+
+func NewMultiSlogHandler(handler slog.Handler) *multiSlogHandler {
```
nit for readability: Comment above constructor
```suggestion
// NewMultiSlogHandler creates an slog.Handler that supports multiple
// underlying handlers, such as to output text and json.
func NewMultiSlogHandler(handler slog.Handler) *multiSlogHandler {
```

This is what the MultiHandler is for right?

joestringer commented on pkg/logging/multihandler.go:
```diff
@@ -0,0 +1,84 @@
+		err := h.Handle(ctx, record)
+		if err != nil {
+			return err
+		}
```
Should this terminate, meaning different handlers receive different output? Or collect errors and attempt to use the later handlers if there's an issue outputting to an earlier handler?

aanm replied:
:bulb: for the collect errors!

joestringer commented on pkg/metrics/logging_hook.go:
```diff
@@ -4,19 +4,29 @@
+
+func init() {
+	// Detector to check if we have duplicate attributes in slog logging hook.
+	slogDupAttrDetection, _ = strconv.ParseBool(os.Getenv("CILIUM_INVALID_METRIC_VALUE_DETECTOR"))
```
nit: Looks like this feature leaked into the second commit, but the third commit seems to be the one intended to introduce the feature.

joestringer reviewed (approved):
LGTM thanks :pray:

tklauser commented on pkg/metrics/logging_hook.go:
```diff
@@ -4,19 +4,29 @@
+
+func init() {
+	// Detector to check if we have duplicate attributes in slog logging hook.
+	slogDupAttrDetection, _ = strconv.ParseBool(os.Getenv("CILIUM_INVALID_METRIC_VALUE_DETECTOR"))
```
Apart from the comment that Joe already made, should this check for `CILIUM_SLOG_DUP_ATTR_DETECTOR` instead of `CILIUM_INVALID_METRIC_VALUE_DETECTOR`?

aanm replied:
Nice catch!

ti-mo reviewed (approved):
Thanks!
